lib/rules/declarations-for-function-components.js

```diff
@@ -0,0 +1,118 @@
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Function declarations for functional components',
+      category: 'Best Practices',
+      recommended: false,
+      url: docsUrl('declarations-for-function-components')
+    },
+
+    schema: []
+  },
+
+  create: Components.detect((context, components) => {
+    const NOT_DECLARED_FUNCTION_COMPONENT_MESSAGE = 'Function component is not a function declaration';
+
+    /**
+     * Mark a component as being a declared function component
+     * @param {ASTNode} node The AST node being checked.
+     */
+    function markComponentAsDeclaredFunctionComponent(node) {
+      components.set(node, {
+        isDeclaredFunctionalComponent: true
+      });
+    }
+
+    /**
+     * Mark a component as being a declared function component
+     * @param {ASTNode} node The AST node being checked.
+     */
+    function markAsMethodDefinition(node) {
+      components.set(node, {
+        isMethodDefinition: true
+      });
+    }
+
+    /**
+     * Mark a component as being a function component
+     * @param {ASTNode} node The AST node being checked.
+     */
+    function markComponentAsFunctionalComponent(node) {
+      components.set(node, {
+        isFunctionalComponent: true
+      });
+    }
+
+    /**
+     * Reports that component is defined with an arrow function
+     * @param {Object} component The component to process
+     */
+    function reportNotDeclaredFunctionalComponent(component) {
+      context.report({
+        node: component.node,
+        message: NOT_DECLARED_FUNCTION_COMPONENT_MESSAGE,
+        data: {
+          component: component.name
+        }
+      });
+    }
+
+    // --------------------------------------------------------------------------
+    // Public
+    // --------------------------------------------------------------------------
+
+    return {
+
+      MethodDefinition(node) {
+        if (components.get(node)) {
+          markAsMethodDefinition(node);
+        }
+      },
+
+      FunctionExpression(node) {
+        if (components.get(node)) {
+          markComponentAsFunctionalComponent(node);
+        }
+      },
+
+      ArrowFunctionExpression(node) {
+        if (components.get(node)) {
+          markComponentAsFunctionalComponent(node);
+        }
+      },
+
+      FunctionDeclaration(node) {
+        if (components.get(node)) {
+          markComponentAsFunctionalComponent(node);
+          markComponentAsDeclaredFunctionComponent(node);
+        }
+      },
+
+      'Program:exit': function () {
+        const list = components.list();
+        // Report that component is arrow function
+        Object.keys(list)
+          .filter(component => list[component].isFunctionalComponent)
+          .filter(component => !list[component].isMethodDefinition)
+          .filter(component => !list[component].isDeclaredFunctionalComponent)
```

Comment:
might as well combine these three filters into one, so it's 3 times more efficient